sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,432 @@
+
+        :param bytes ciphertext: Encrypted bytes to decrypt.
+        :param padding: The padding to use. Supported paddings are `OAEP` and `PKCS1v15`. For `OAEP` padding, supported
+            hash algorithms are `SHA1` and `SHA256`. The only supported mask generation function is `MGF1`. See
+            https://learn.microsoft.com/azure/key-vault/keys/about-keys-details for details.
+        :type padding: :class:`~cryptography.hazmat.primitives.asymmetric.padding.AsymmetricPadding`
+
+        :returns: The decrypted plaintext, as bytes.
+        :rtype: bytes
+        """
+        mapped_algorithm = get_encryption_algorithm(padding)
+        result = self._client.decrypt(mapped_algorithm, ciphertext)
+        return result.plaintext
+
+    @property
+    def key_size(self) -> int:
+        """The bit length of the public modulus.
+
+        :returns: The key's size.
+        :rtype: int
+        """
+        private_key = self.private_numbers().private_key()
+        return private_key.key_size
+
+    def public_key(self) -> KeyVaultRSAPublicKey:
+        """The `RSAPublicKey` associated with this private key, as a `KeyVaultRSAPublicKey`.
+
+        The public key implementation will use the same underlying cryptography client as this private key.
+
+        :returns: The `KeyVaultRSAPublicKey` associated with the key.
+        :rtype: :class:`~azure.keyvault.keys.crypto.KeyVaultRSAPublicKey`
+        """
+        return KeyVaultRSAPublicKey(self._client, self._key)
+
+    def sign(
+        self,
+        data: bytes,
+        padding: AsymmetricPadding,
+        algorithm: Union[Prehashed, HashAlgorithm],
+    ) -> bytes:
+        """Signs the data.
+
+        :param bytes data: The data to sign, as bytes.
+        :param padding: The padding to use. Supported paddings are `PKCS1v15` and `PSS`. For `PSS`, the only supported
+            mask generation function is `MGF1`. See https://learn.microsoft.com/azure/key-vault/keys/about-keys-details
+            for details.
+        :type padding: :class:`~cryptography.hazmat.primitives.asymmetric.padding.AsymmetricPadding`
+        :param algorithm: The algorithm to sign with. Only `HashAlgorithm`s are supported -- specifically, `SHA256`,
+            `SHA384`, and `SHA512`.
+        :type algorithm: :class:`~cryptography.hazmat.primitives.asymmetric.utils.Prehashed` or
+            :class:`~cryptography.hazmat.primitives.hashes.HashAlgorithm`
+
+        :returns: The signature, as bytes.
+        :rtype: bytes
+        """
+        if isinstance(algorithm, Prehashed):
+            raise ValueError("`Prehashed` algorithms are unsupported. Please provide a `HashAlgorithm` instead.")
+        mapped_algorithm = get_signature_algorithm(padding, algorithm)
+        digest = Hash(algorithm)
+        digest.update(data)
+        result = self._client.sign(mapped_algorithm, digest.finalize())
+        return result.signature
+
+    def private_numbers(self) -> RSAPrivateNumbers:
```

Comment:
I don't think the private key numbers will ever be returned from the service, does this take that into account?

Reply:
This does, yes; we raise an error if we don't have the necessary information (we need at least `d` to calculate the rest) and accounts for private numbers possibly being absent.

We can use a private key, but only by creating a crypto client with a local JWK that has the necessary information (we do this in `test_crypto_client.py` to verify this method and `private_bytes`). You're right that using a key that's fetched from the service won't have the private material returned.